### DISCUSSION TITLE
Posix UUIDs are larger than `utmp`/`wtmp` allows

### DISCUSSION THREAD

**Discussion by TheRealGramdalf** [kanidm/kanidm#2826] [General]
### I did this

Attempt to modify the NixOS module for users to allow defining user configuration without actually creating the user

### I expected the following

Everything should work first try... right?

### This happened instead

When attempting to add a user from Kanidm by UUID, NixOS throws:
```
Username '9c7d73a0-ef7e-45bc-a314-947ec6ab00e6' is longer than 31 characters which is not allowed!
```

### Kanidm version details

N/A

### Any other comments

Upon further inspection:
it stems from [this PR](https://github.com/NixOS/nixpkgs/pull/36556), which was added as a check to help prevent misconfiguration according to systemds' limits;
which is implemented [here](https://github.com/systemd/systemd/blob/73df6cd3ee4ac88953a91cd20707f7f5b8fa7d21/src/basic/user-util.c#L804-L805);
and the rationale explained [here](https://github.com/systemd/systemd/blob/73df6cd3ee4ac88953a91cd20707f7f5b8fa7d21/docs/USER_NAMES.md?plain=1#L82-L87).

The limiting factor is UTMP, which has a practical length limit of `31`; far smaller than the other limitations ranging 250+. Even if the dashes were removed, Kanidms' UUIDs are still one character larger than `utmp` allows. This makes using a kanidm user in a NixOS configuration much harder, since usernames are unstable. 

Since the display name should be used for `utmp`/`wtmp` in most cases, is it safe to simply add an override switch that disables the check (and use systemd's "relaxed" mode, which does not apply character limits (https://github.com/systemd/systemd/blob/73df6cd3ee4ac88953a91cd20707f7f5b8fa7d21/src/basic/user-util.c#L713), or does more drastic action (such as patching `utmp/wtmp` need to take place?

**Comment by yaleman** [comment]
This isn't an issue relating to the design or requiring action from the Kanidm project, so it's been moved here. Even then, why's this not in a NixOS-related space, since it's related to a platform-specific thing?

**Reply by Firstyear** [reply]
If I recall utmp isn't compatible with 2038 problem so it's getting removed in other distros: https://microos.opensuse.org/blog/2023-09-28-replace-utmp-with-logind/

**Reply by TheRealGramdalf** [reply]
I suppose it is more of a platform specific thing than I thought. I mainly opened a bug report because it looked as if there would be issues with `kanidm-unixd`: since `utmp` and `wtmp` (which to my understanding handles things like names in logfiles, e.g. `Jun 06 19:07:01 aerwiar nix-daemon[2015]: accepted connection from pid 150147, user gramdalf`), would there not be issues if `uid_attr_map` was set to `uuid` (or an SPN that was 31 or more characters)?

I suppose the way I worded it is mainly focusing on NixOS, but from what I can tell (which could be completely wrong), there may be an impact on `kanidm-unixd` if usernames are too long:

> ...`UT_NAMESIZE-1` (typically 31 on Linux; rationale: names longer than
this cannot correctly appear in `utmp`/`wtmp` and create ambiguity with login
accounting)...

among other reasons mentioned in the links above.

**Reply by TheRealGramdalf** [reply]
> If I recall utmp isn't compatible with 2038 problem so it's getting removed in other distros: https://microos.opensuse.org/blog/2023-09-28-replace-utmp-with-logind/

That may cut the problems out entirely - if there's already a better replacement I'm happy to just use that. I haven't looked that far into logind/utmp (hence the uncertainty in the post/other reply) - it's on my list of things to double check configuration for.
